File: python/ssd_caffe.py
#coding=utf-8
# 极速检测器
# 烟雨@烟雨科技
# V2.0 2020.11.28
import sys
import os
import time
import platform
import argparse
import random
import numpy as np
import cv2
if platform.system()=="Windows":
    caffe_root = 'D:/CNN/ssd'
else:
    caffe_root = os.path.expanduser('~') + "/CNN/ssd"
sys.path.insert(0, caffe_root + '/python')
import caffe

rootdir = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))

CLASSES = ('aeroplane', 'bicycle', 'bird', 'boat','bottle', 'bus', 'car', 'cat', 'chair','cow', 'diningtable', 'dog', 'horse','motorbike', 'person', 'pottedplant','sheep', 'sofa', 'train', 'tvmonitor')

def drawDetection(image, detections, cost = None, usecolor = None):
    if image is None:
        return image
    height, width, _ = image.shape
    for item in detections:
        xmin = int(round(item[3] * width))
        ymin = int(round(item[4] * height))
        xmax = int(round(item[5] * width))
        ymax = int(round(item[6] * height))
        label = CLASSES[int(item[1])-1]
        if usecolor is None:
            cv2.putText(image,label,(xmin,ymin), 3,1,(255,0,0))
            cv2.rectangle(image,(xmin,ymin),(xmax,ymax),(255,0,0))
        else:
            import seaborn
            colors = seaborn.color_palette("hls", len(CLASSES))
            color = colors[int(round(item[1]))]
            color = [c *256 for c in color]
            cv2.putText(image,label,(xmin,ymin), 3,1,color)
            cv2.rectangle(image,(xmin,ymin),(xmax,ymax),color)

    if not cost is None:
        cv2.putText(image,cost,(0,40),3,1,(0,0,255))

    return image

class MobileNetSSDDetector:
    def __init__(self,model_def, model_weights):
        if not os.path.exists(model_weights):
            print(model_weights + "does not exist,")
            print("use merge_bn.py to generate it.")
            exit()
        self.net = caffe.Net(model_def,model_weights,caffe.TEST)
        self.net_width = self.net.blobs[self.net.inputs[0]].width
        self.net_height = self.net.blobs[self.net.inputs[0]].height

    def preprocess(self,src):
        img = cv2.resize(src, (self.net_width, self.net_height))
        img = (img - 127.5) * 0.007843
        return img

    def detectAndDraw(self,img):
        start = time.time()
        result = self.detect(img)
        end = time.time()
        cost="%0.2fms" %((end-start)*1000)
        show = drawDetection(img,result, cost)
        return show

    def detect(self,img):
        self.net.blobs['data'].data[...] = self.preprocess(img).transpose((2, 0, 1)) 
        #start = time.time()
        detections = self.net.forward()[self.net.outputs[0]][0][0]
        #end = time.time()
        #cost = "det cost: %0.2fms" %((end-start)*1000)
        #print(cost)
        return detections

def testdir(detector, dir = "images"):
    files = os.listdir(dir)
    for file in files:
        imgfile = dir + "/" + file
        img = cv2.imread(imgfile)
        show = detector.detectAndDraw(img)
        cv2.imshow("img",show)
        cv2.waitKey()

def testcamera(detector):
    cap = cv2.VideoCapture(0)
    while True:
        ret, img = cap.read()
        if not ret:
            break
        show = detector.detectAndDraw(img)
        cv2.imshow("img",show)
        cv2.waitKey(1)

def testvideos(detector,dir="videos"):
    files = os.listdir(dir)
    for file in files:
        filepath = dir + "/" + file
        cap = cv2.VideoCapture(filepath)
        while True:
            ret, img = cap.read()
            if not ret:
                break
            show = detector.detectAndDraw(img)
            cv2.imshow("img", show)
            cv2.waitKey(1)

def main(args):
    model_def = rootdir+"/"+args.dataset+'/MobileNetSSD_deploy_depth.prototxt'
    model_weights = rootdir+"/"+args.dataset+'/MobileNetSSD_deploy.caffemodel'
    if not os.path.exists(model_weights):
        print(model_weights+" does not exist,")
        print("use merge_bn.py to generate it.")
        exit()
    if not platform.system() == "Darwin":
        caffe.set_device(0)
        caffe.set_mode_gpu()
    detector = MobileNetSSDDetector(model_def, model_weights)   
    #testcamera(detector)
    #testvideos(detector,args.video_dir)
    testdir(detector)

def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('--dataset', default="Face")
    return parser.parse_args()

if __name__ == '__main__':
    args = parse_args()
    main(args)

File: cpp/ssd_caffe.cpp
#include <caffe/caffe.hpp>
#include "mrdir.h"
#include "mropencv.h"
#include <algorithm>
#include <iomanip>
#include <iosfwd>
#include <memory>
#include <string>
#include <utility>
#include <vector>

using namespace caffe;  // NOLINT(build/namespaces)

class Detector {
public:
	Detector(const string& model_file,const string& weights_file);

	std::vector<vector<float> > Detect(const cv::Mat& img);

private:
	void WrapInputLayer(std::vector<cv::Mat>* input_channels);

	void Preprocess(const cv::Mat& img,
		std::vector<cv::Mat>* input_channels);

private:
	shared_ptr<Net<float> > net_;
	cv::Size input_geometry_;
	int num_channels_;
};

Detector::Detector(const string& model_file,
	const string& weights_file) {
#ifdef CPU_ONLY
	Caffe::set_mode(Caffe::CPU);
#else
	Caffe::set_mode(Caffe::GPU);
#endif

	/* Load the network. */
	net_.reset(new Net<float>(model_file, TEST));
	net_->CopyTrainedLayersFrom(weights_file);

	CHECK_EQ(net_->num_inputs(), 1) << "Network should have exactly one input.";
	CHECK_EQ(net_->num_outputs(), 1) << "Network should have exactly one output.";

	Blob<float>* input_layer = net_->input_blobs()[0];
	num_channels_ = input_layer->channels();
	CHECK(num_channels_ == 3 || num_channels_ == 1)
		<< "Input layer should have 1 or 3 channels.";
	input_geometry_ = cv::Size(input_layer->width(), input_layer->height());
}

std::vector<vector<float> > Detector::Detect(const cv::Mat& img) {
	Blob<float>* input_layer = net_->input_blobs()[0];
	input_layer->Reshape(1, num_channels_,
		input_geometry_.height, input_geometry_.width);
	/* Forward dimension change to all layers. */
	net_->Reshape();

	std::vector<cv::Mat> input_channels;
	WrapInputLayer(&input_channels);

	Preprocess(img, &input_channels);

	net_->Forward();

	/* Copy the output layer to a std::vector */
	Blob<float>* result_blob = net_->output_blobs()[0];
	const float* result = result_blob->cpu_data();
	const int num_det = result_blob->height();
	vector<vector<float> > detections;
	for (int k = 0; k < num_det; ++k) {
		if (result[0] == -1) {
			// Skip invalid detection.
			result += 7;
			continue;
		}
		vector<float> detection(result, result + 7);
		detections.push_back(detection);
		result += 7;
	}
	return detections;
}

/* Wrap the input layer of the network in separate cv::Mat objects
* (one per channel). This way we save one memcpy operation and we
* don't need to rely on cudaMemcpy2D. The last preprocessing
* operation will write the separate channels directly to the input
* layer. */
void Detector::WrapInputLayer(std::vector<cv::Mat>* input_channels) {
	Blob<float>* input_layer = net_->input_blobs()[0];

	int width = input_layer->width();
	int height = input_layer->height();
	float* input_data = input_layer->mutable_cpu_data();
	for (int i = 0; i < input_layer->channels(); ++i) {
		cv::Mat channel(height, width, CV_32FC1, input_data);
		input_channels->push_back(channel);
		input_data += width * height;
	}
}

void Detector::Preprocess(const cv::Mat& img,
	std::vector<cv::Mat>* input_channels) {
	/* Convert the input image to the input image format of the network. */
	cv::Mat sample;
	if (img.channels() == 3 && num_channels_ == 1)
		cv::cvtColor(img, sample, cv::COLOR_BGR2GRAY);
	else if (img.channels() == 4 && num_channels_ == 1)
		cv::cvtColor(img, sample, cv::COLOR_BGRA2GRAY);
	else if (img.channels() == 4 && num_channels_ == 3)
		cv::cvtColor(img, sample, cv::COLOR_BGRA2BGR);
	else if (img.channels() == 1 && num_channels_ == 3)
		cv::cvtColor(img, sample, cv::COLOR_GRAY2BGR);
	else
		sample = img;

	cv::Mat sample_resized;
	if (sample.size() != input_geometry_)
		cv::resize(sample, sample_resized, input_geometry_);
	else
		sample_resized = sample;

	cv::Mat sample_float;
	if (num_channels_ == 3)
		sample_resized.convertTo(sample_float, CV_32FC3);
	else
		sample_resized.convertTo(sample_float, CV_32FC1);

	cv::Mat sample_normalized;
//	cv::subtract(sample_float, mean_, sample_normalized);
	sample_normalized = (sample_float - 127.5)*0.007843;
	/* This operation will write the separate BGR planes directly to the
	* input layer of the network because it is wrapped by the cv::Mat
	* objects in input_channels. */
	cv::split(sample_normalized, *input_channels);

	CHECK(reinterpret_cast<float*>(input_channels->at(0).data)
		== net_->input_blobs()[0]->cpu_data())
		<< "Input channels are not wrapping the input layer of the network.";
}
DEFINE_double(confidence_threshold, 0.9,
	"Only store detections with score higher than the threshold.");
#include "fstream"
static void split(const std::string& s, const std::string delim, std::vector< std::string >* ret)
{
	size_t last = 0;
	size_t index = s.find_first_of(delim, last);
	while (index != std::string::npos)
	{
		ret->push_back(s.substr(last, index - last));
		last = index + 1;
		index = s.find_first_of(delim, last);
	}
	if (index - last > 0)
	{
		ret->push_back(s.substr(last, index - last));
	}
}
vector<string>getLabels(const string labeltxt)
{
	vector<string>labels;
	std::ifstream labelmapfile(labeltxt);
	string line;
	while (getline(labelmapfile, line))
	{
		vector<string>strs;
		split(line, ":", &strs);
		if (strs.size() == 2 && strs[0] == "  name")
		{
			string label = strs[1];
			label = label.substr(2, label.length() - 3);
			labels.push_back(label);
		}
	}
	return labels;
}

cv::Mat ShowDetectionResult(const cv::Mat &img, const std::vector<vector<float> > &detections, std::vector<std::string> &labels, double confidence_threshold)
{
	cv::Mat show=img.clone();
	for (int i = 0; i < detections.size(); ++i)
	{
		const vector<float>& d = detections[i];
		// Detection format: [image_id, label, score, xmin, ymin, xmax, ymax].
		CHECK_EQ(d.size(), 7);
		const float score = d[2];
		if (score >= confidence_threshold)
		{
			cv::rectangle(show, cv::Point(d[3] * img.cols, d[4] * img.rows), cv::Point(d[5] * img.cols, d[6] * img.rows), CV_RGB(255, 0, 0));
			cv::putText(show, labels[static_cast<int>(d[1])], cv::Point(d[3] * img.cols, d[4] * img.rows), 2, 1, CV_RGB(255, 0, 0));
		}
	}
	return show;
}

void testdir(Detector &detector, std::vector<std::string> &labels,double confidence_threshold,const string dir = "../images")
{
	auto files = getAllFilesinDir(dir);
	for (int i = 0; i < files.size(); i++)
	{
		std::string filepath = dir + "/" + files[i];
		cv::Mat img = cv::imread(filepath, -1);
		std::cout << files[i] << std::endl;
		CHECK(!img.empty()) << "Unable to decode image " << filepath;
		cv::TickMeter tm;
		tm.start();
		std::vector<vector<float> > detections = detector.Detect(img);
		tm.stop();
		std::cout << tm.getTimeMilli() << "ms" << std::endl;
		/* Print the detection results. */
		cv::Mat show = ShowDetectionResult(img, detections,labels,confidence_threshold);
		cv::imshow("img", show);
		cv::waitKey();
		}
	cv::waitKey();
}

void testcamera(Detector &detector, std::vector<std::string> &labels, double confidence_threshold, int cameraindex = 0)
{
	cv::VideoCapture cap(cameraindex);
	if (!cap.isOpened()) {
		LOG(FATAL) << "Failed to open camera ";
	}
	cv::Mat img;
	int frame_count = 0;
	while (true) {
		bool success = cap.read(img);
		if (!success) {
			LOG(INFO) << "Process " << frame_count << " frames";
			break;
		}
		CHECK(!img.empty()) << "Error when read frame";
		cv::TickMeter tm;
		tm.start();
		std::vector<vector<float> > detections = detector.Detect(img);
		tm.stop();
		std::cout << tm.getTimeMilli() << "ms"<<std::endl;
		cv::Mat show = ShowDetectionResult(img, detections, labels, confidence_threshold);
		++frame_count;
		cv::imshow("img", show);
		cv::waitKey(1);
	}
	if (cap.isOpened()) {
		cap.release();
	}
}

int main(int argc, char** argv) {
	::google::InitGoogleLogging(argv[0]);
	// Print output to stderr (while still logging)
	FLAGS_alsologtostderr = 1;

#ifndef GFLAGS_GFLAGS_H_
	namespace gflags = google;
#endif

	gflags::SetUsageMessage("Do detection using SSD mode.\n"
		"Usage:\n"
		"    ssd_detect [FLAGS] model_file weights_file list_file\n");
	gflags::ParseCommandLineFlags(&argc, &argv, true);
	const string& dataset = "Face";
	const string& labelfilepath = "../" + dataset + "/labelmap.prototxt";
	const string& model_file = "../"+dataset+"/MobileNetSSD_deploy_depth.prototxt";
	const string& weights_file = "../" + dataset + "/MobileNetSSD_face.caffemodel";
	const float confidence_threshold = FLAGS_confidence_threshold;
	// Initialize the network.
	Detector detector(model_file, weights_file);
	auto labels = getLabels(labelfilepath);
	testdir(detector, labels, confidence_threshold);
	//testcamera(detector, labels, confidence_threshold);
	return 0;
}